src/session_pool/session_pool.js

```diff
@@ -0,0 +1,314 @@
+     * @return {Promise<Session>}
+     */
+    async retrieveSession() {
+        // If we have enough space for session. Return newly created session.
+        if (this._hasSpaceForSession()) {
+            return this._createSession();
+        }
+
+        // For example that developer can plug different picking algorithms such as the Lukášův
+        // Maybe a should have pickSession function to be customizable.
+        const pickedSession = this._pickSession();
+
+        // If session can be used return the session
+        if (pickedSession.isUsable()) {
+            return pickedSession;
+        }
+
+        //  otherwise remove old session and return newly created session
+        this._removeSession(pickedSession);
+        return this._createSession();
+    }
+
+    /**
+     * Returns an object representing the internal state of the `SessionPool` instance.
+     * Note that the object's fields can change in future releases.
+     *
+     * @returns {Object}
+     */
+    getState() {
+        return {
+            usableSessionsCount: this.usableSessionsCount,
+            retiredSessionsCount: this.retiredSessionsCount,
+            sessions: this.sessions.map(session => session.getState()),
+        };
+    }
+
+    /**
+     * Persists the current state of the `SessionPool` into the default {@link KeyValueStore}.
+     * The state is persisted automatically in regular intervals.
+     *
+     * @return {Promise}
+     */
+    async persistState() {
+        log.debug('SessionPool: Persisting state');
+        log.debug(`SessionPool: persistStateKeyValueStoreId: ${this.persistStateKeyValueStoreId}, persistStateKey: ${this.persistStateKey} `);
+        return this.keyValueStore.setValue(this.persistStateKey, this.getState());
+    }
+
+    /**
+     * Removes `Session` instance from `SessionPool`.
+     * @param session {Session} - Session to be removed
+     * @private
+     */
+    _removeSession(session) {
+        const sessionIndex = this.sessions.findIndex(storedSession => storedSession.id === session.id);
+
+        const removedSession = this.sessions.splice(sessionIndex, 1);
+        log.debug(`SessionPool: Removed Session - ${removedSession.id}`);
+    }
+
+    /**
+     * Adds `Session` instance to `SessionPool`.
+     * @param newSession {Session} - `Session` instance to be added.
+     * @private
+     */
+    _addSession(newSession) {
+        this.sessions.push(newSession);
+    }
+
+    /**
+     * Gets random index.
+     * @return {number}
+     * @private
+     */
+    _getRandomIndex() {
+        return Math.floor(Math.random() * this.sessions.length);
+    }
+
+    /**
+     * Creates new session without any extra behavior.
+     * @return {Session} - New session.
+     * @private
+     */
+    _defaultCreateSessionFunction() {
+        return new Session({
+            maxSessionAgeSecs: this.maxSessionAgeSecs,
+            maxSessionUsageCount: this.maxSessionUsageCount,
+            sessionPool: this,
```

Comment:
How am I going to pass/closure `sessionPool` to my custom `createSessionFunction`, which I need to define in the `SessionPool` constructor, where the `sessionPool` doesn't exist yet?

I think the custom `createSessionFunction` needs to be wrapped by a private function that injects the `sessionPool`.